keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1266 @@
+  // the backoff limit per index.
+  // +optional
+  RunPolicy *JobRunPolicy
+  ...
+}
+
+type JobStatus struct {
+  ...
+
+  // FailedIndexes holds the failed indexes when backoffLimitPerIndex is set.
+  // The indexes are represented in the text format analogous as for the
+  // `completedIndexes` field, ie. they are kept as decimal integers
+  // separated by commas. The numbers are listed in increasing order. Three or
+  // more consecutive numbers are compressed and represented by the first and
+  // last element of the series, separated by a hyphen.
+  // For example, if the failed indexes are 1, 3, 4, 5 and 7, they are
+  // represented as "1,3-5,7".
+  // +optional
+  FailedIndexes *string
+}
+```
+
+Here, setting `.spec.runPolicy.backoffLimitPerIndex` along with `.spec.backoffLimit`
+results in a warning. Still, we leave it possible to set both fields to allow
+for the use case of a more controller downgrade.
+
+### Tracking the number of failures per index
+
+In order to determine if the backoff limit per index is exceeded we keep
+track of the number of failures per index. For this purpose we use the Pod
+annotation, `batch.kubernetes.io/job-index-failure-count`, which holds the value
+of the number of pod failures for a given index. It is set to `0` for the first
+pod created for a given index.
+
+When Job controller sees a failed pod corresponding to a given index, and the
+value of the annotation `batch.kubernetes.io/job-index-failure-count` is greater
+or equal to the configured backoff limit per index then the index is marked
+as failed and added to `.status.failedIndexes`.
+
+When Job controller creates replacement pods for failed pods for a given
+index it checks if the index isn't finished yet (it is not in
+`.status.failedIndexes` nor `.status.completedIndexes`).
+Then, if `x` is the highest `batch.kubernetes.io/job-index-failure-count`
+for the index, the newly created pod will have the annotation set to `x+1`.
+
+In order to keep track of the number of failures per index, the Job controller
+removes finalizers of a failed pod for a given index, only once the replacement
+pod (with incremented value of `batch.kubernetes.io/job-index-failure-count`) is
+created, or the index is marked as failed in `.status.failedIndexes`. This means
+that these are the main steps when handling a failed pod to prepare it for
+deletion:
+1. Pod is recognized as failed
+2. pod UID is recorded in Job status (`.status.uncountedTerminatedPods`)
+3. the replacement Pod is created
+4. Pod's finalizer is removed
+
+Here, the new feature adds a dependency between steps (3.) and (4.) as previously
+these steps could be performed in any order.
+
+Note that, the first step (1.) will also be impacted by
+[KEP-3939: Consider Terminating pods as active pods in Jobs.](https://github.com/kubernetes/enhancements/issues/3939)
+
+### Failed indexes format
+
+The format of the `.status.failedIndexes` field is analogous to the one used for
```

Comment:
Maybe link to https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2214-indexed-job#track-completed-indexes-in-job-status